### DISCUSSION TITLE
Anyone integrated with OpenID Connect provider?

### DISCUSSION THREAD

**Discussion by candlerb** [netbox-community/netbox#8579] [General]
Does anyone have any experience to share with integrating with a local OpenID Connect provider using the python-social-auth stuff in Netbox 3.1? Documentation appears to be minimal.

I have a local OIDC provider (Hashicorp Vault).  After going through some python-social-auth source code, I reckoned I would need something like this:

```
REMOTE_AUTH_BACKEND = 'social_core.backends.open_id_connect.OpenIdConnectAuth'
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_DEFAULT_GROUPS = []
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
# Not sure what name this setting should have
SOCIAL_AUTH_OIDC_ENDPOINT = 'https://vault.example.net:8200/v1/identity/oidc/provider/default' 
# and I'll need to add client ID and client secret somewhere
```

First problem:

```
<class 'ModuleNotFoundError'>

No module named 'jose'

Python version: 3.8.10
NetBox version: 3.1.7
```

I tried installing `jose` inside the venv but that was a mistake; it is ancient python2 code. So I uninstalled `jose` and installed `python-jose[cryptography]` instead.

Now I can get to the Netbox home page.  But when I click on the "Login" button:

```
<class 'django.urls.exceptions.NoReverseMatch'>

Reverse for 'begin' with keyword arguments '{'backend': ''}' not found. 1 pattern(s) tried: ['oauth/login/(?P<backend>[^/]+)/\\Z']

Python version: 3.8.10
NetBox version: 3.1.7
```

Backtrace:

```
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/decorators/debug.py", line 89, in sensitive_post_parameters_wrapper
    return view(request, *args, **kwargs)
  File "/opt/netbox/netbox/users/views.py", line 38, in dispatch
    return super().dispatch(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 98, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox/netbox/users/views.py", line 47, in get
    return render(request, self.template_name, {
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/shortcuts.py", line 19, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/loader.py", line 62, in render_to_string
    return template.render(context, request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 170, in render
    return self._render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/test/utils.py", line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/test/utils.py", line 100, in instrumented_test_render
    return self.nodelist.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/defaulttags.py", line 315, in render
    return nodelist.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 938, in render
    bit = node.render_annotated(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/defaulttags.py", line 214, in render
    nodelist.append(node.render_annotated(context))
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/template/defaulttags.py", line 449, in render
    url = reverse(view_name, args=args, kwargs=kwargs, current_app=current_app)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/urls/base.py", line 86, in reverse
    return resolver._reverse_with_prefix(view, prefix, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/urls/resolvers.py", line 698, in _reverse_with_prefix
    raise NoReverseMatch(msg)

Exception Type: NoReverseMatch at /login/
Exception Value: Reverse for 'begin' with keyword arguments '{'backend': ''}' not found. 1 pattern(s) tried: ['oauth/login/(?P<backend>[^/]+)/\\Z']
```

Any clues?

**Comment by candlerb** [comment]
I finally got this to work, but I had to subclass [open_id_connect.py](https://github.com/python-social-auth/social-core/blob/master/social_core/backends/open_id_connect.py) to set a backend name and override some settings to work with Vault.

1. Add to `local_requirements.txt`:
    ```
    python-jose[cryptography]
    ```
    and install it (i.e. activate the virtualenv then `pip install -r local_requirements.txt`)
2. Apply [this patch](https://github.com/python-social-auth/social-core/commit/36f25c4218f17c1c74fbb22881716d3c4bdbd379) which was committed after social-core 4.2.0
3. Install [this file](https://gist.github.com/candlerb/6702be8aa19cbe5762b7c84f6f584e29) as `/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/vault.py`
4. Create client on upstream Vault OIDC provider with redirect_uri `https://netbox.your-domain.com/oauth/complete/vault/` (including trailing slash)
5. Configure Netbox:
    ```
    REMOTE_AUTH_ENABLED = True
    REMOTE_AUTH_BACKEND = 'social_core.backends.vault.VaultOpenIdConnect'
    ...
    SOCIAL_AUTH_VAULT_OIDC_ENDPOINT = 'https://vault.example.net:8200/v1/identity/oidc/provider/<provider-name>'
    SOCIAL_AUTH_VAULT_KEY = '<client_id>'
    SOCIAL_AUTH_VAULT_SECRET = '<client_secret>'
    SOCIAL_AUTH_VAULT_SCOPE = ['groups']    # optional, since returned groups aren't used
    ```
6. Restart Netbox

I don't have group sync working, and I can only see group sync code in Netbox for the `RemoteUserBackend` and possibly `LDAPBackend`.  However I can set group membership locally within Netbox, and set staff / superuser flags locally.

Only two other existing social-auth backends return groups: [mediawiki](https://github.com/python-social-auth/social-core/blob/master/social_core/backends/mediawiki.py) and [discourse](https://github.com/python-social-auth/social-core/blob/master/social_core/backends/discourse.py).  If I try to return groups like this:

```
    def get_user_details(self, response):
        username_key = self.setting('USERNAME_KEY', default=self.USERNAME_KEY)
        return {
            ...
            'groups': response.get('groups'),
        }
```

then I get a Django ORM exception:

```
Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_django/utils.py", line 46, in wrapper
    return func(request, backend, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_django/views.py", line 31, in complete
    return do_complete(request.backend, _do_login, user=request.user,
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/actions.py", line 45, in do_complete
    user = backend.complete(user=user, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/base.py", line 40, in complete
    return self.auth_complete(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/utils.py", line 247, in wrapper
    return func(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/oauth.py", line 391, in auth_complete
    return self.do_auth(response['access_token'], response=response,
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/utils.py", line 247, in wrapper
    return func(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/oauth.py", line 403, in do_auth
    return self.strategy.authenticate(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_django/strategy.py", line 105, in authenticate
    return authenticate(*args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/decorators/debug.py", line 42, in sensitive_variables_wrapper
    return func(*func_args, **func_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/contrib/auth/__init__.py", line 76, in authenticate
    user = backend.authenticate(request, **credentials)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/base.py", line 80, in authenticate
    return self.pipeline(pipeline, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/base.py", line 83, in pipeline
    out = self.run_pipeline(pipeline, pipeline_index, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/backends/base.py", line 113, in run_pipeline
    result = func(*args, **out) or {}
  File "/opt/netbox/venv/lib/python3.8/site-packages/social_core/pipeline/user.py", line 119, in user_details
    setattr(user, name, value)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 545, in __set__
    raise TypeError(

Exception Type: TypeError at /oauth/complete/vault/
Exception Value: Direct assignment to the forward side of a many-to-many set is prohibited. Use groups.set() instead.
```

This suggests to me that it's trying to assign a list of groups directly to the `groups` attribute of a `User` object, which is not permitted (as the error says). I guess it's possible to modify Netbox or the social-auth pipeline to deal with this, but I don't really have the bandwidth to dig any further right now.

**Comment by empusas** [comment]
I am having also the same issue using our company SSO with netbox. Can you elaborate a bit more how you fixed the URL error? 

```
REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'social_core.backends.open_id_connect.OpenIdConnectAuth'
REMOTE_AUTH_AUTO_CREATE_USER = True
SOCIAL_AUTH_OPEN_ID_CONNECT_KEY='secret'
SOCIAL_AUTH_OPEN_ID_CONNECT_SECRECT='secret'
SOCIAL_AUTH_OPEN_ID_CONNECT_API_URL='https://test-login.abc.net/oidc/endpoint/default/authorize/'
SOCIAL_AUTH_PIPELINE = [
    "social_core.pipeline.social_auth.social_details",
    "social_core.pipeline.social_auth.social_uid",
    "social_core.pipeline.social_auth.auth_allowed",
    "social_core.pipeline.social_auth.social_user",
    "social_core.pipeline.social_auth.associate_by_email",
    "social_core.pipeline.user.create_user",
    "social_core.pipeline.social_auth.associate_user",
    "social_core.pipeline.social_auth.load_extra_data",
    "social_core.pipeline.user.user_details"
]
```
 
This is the error I got:
 
 
```
<class 'django.urls.exceptions.NoReverseMatch'>
 
Reverse for 'begin' with keyword arguments '{'backend': ''}' not found. 1 pattern(s) tried: ['oauth/login/(?P<backend>[^/]+)/\\Z']
 
Python version: 3.9.9
NetBox version: 3.1.5
```

**Reply by candlerb** [reply]
It's quite a long story if you want to use python-social-auth :-)

Firstly, the `open_id_connect.OpenIdCOnnectAuth` provider *cannot* be instantiated as-is.  It's a base class, which has to be subclassed.

You can either make your own backend which subclasses it, as I did in pull request [#668](https://github.com/python-social-auth/social-core/pull/668) to integrate with Vault; or you can apply this pull request [#669](https://github.com/python-social-auth/social-core/pull/669) which makes it generic, so that you *can* instantiate `OpenIdConnectAuth` directly.  Of course, if you're running a patched version of python-social-auth then you're on your own if it breaks.

Now, I'm not sure about your NoReverseMatch error, but it could be that you've configured the wrong redirect_uri in your provider. It should be `https://netbox.example.com/oauth/oidc/complete/` (with the trailing slash); replace `oidc` with your subclass name if you made a subclass.  (EDIT: I think this is because you tried to use `social_core.backends.open_id_connect.OpenIdConnectAuth` directly, but it has no name. This is one of several reasons why it can't be instantiated directly.)

Since I'm using Apache as the front-end to Netbox anyway, I have changed back to using `mod_auth_openidc`.  This works well, except it doesn't give you a login page that gives you the choice between local username/password and social auth provider. It does however have the major advantage that it can sync groups to Netbox, which doesn't work with Django + python-social-auth.

Netbox configuration:

```
REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'netbox.authentication.RemoteUserBackend'
REMOTE_AUTH_HEADER = 'HTTP_REMOTE_USER'
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_DEFAULT_GROUPS = []
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
REMOTE_AUTH_GROUP_HEADER = 'HTTP_OIDC_CLAIM_GROUPS'
REMOTE_AUTH_GROUP_SYNC_ENABLED = True
REMOTE_AUTH_GROUP_SEPARATOR = ','
REMOTE_AUTH_STAFF_GROUPS = ['netbox:staff']
REMOTE_AUTH_SUPERUSER_GROUPS = ['netbox:superuser']
```

`mod_auth_openidc` config, inserted inside the netbox virtualhost:

```
    OIDCProviderMetadataURL https://XXXXXXXX/.well-known/openid-configuration
    OIDCClientID XXXXXXXX
    OIDCClientSecret XXXXXXXX

    OIDCRedirectURI /oauth2callback
    OIDCCryptoPassphrase XXXXXXXX
    OIDCScope "openid email profile groups"
    OIDCSessionInactivityTimeout 43200
    OIDCCookieSameSite On
    #Only available in mod_auth_openidc 2.3.8+:
    OIDCStateMaxNumberOfCookies 7 true

    # Note that OIDC Providers (OP) don't guarantee the uniqueness of these claims
    #OIDCRemoteUserClaim email
    OIDCRemoteUserClaim preferred_username
    OIDCPassUserInfoAs claims
    OIDCPassClaimsAs headers
    RequestHeader unset REMOTE_USER
    RequestHeader set REMOTE_USER expr=%{REMOTE_USER}

    <Location />
        AuthType openid-connect
        Require claim groups:netbox:access     # or: Require valid-user
        AuthzSendForbiddenOnFailure on
    </Location>

    <Location /api/>
        <RequireAny>
            AuthType openid-connect
            Require claim groups:netbox:access     # or: Require valid-user
            # Allow access to /api/ with token instead of OIDC:
            # https://httpd.apache.org/docs/2.4/howto/access.html#env
            Require expr "%{HTTP:Authorization} =~ /^Token /"
        </RequireAny>
        AuthzSendForbiddenOnFailure on
    </Location>
```

With this configuration the redirect_uri at the IDP is `https://netbox.example.com/oauth2callback`

If you're using Nginx as the front-end then you might want to switch to Apache, because as far as I can tell, you need the commercial version of Nginx to do OIDC.

**Comment by jschewebbn** [comment]
Did you ever figure out group sync support?

**Reply by rakesh-p** [reply]
@candlerb  after you guided me in #12349, i also tried python-social-auth with keycloak backend (REMOTE_AUTH_BACKEND = 'social_core.backends.keycloak.KeycloakOAuth2'). Log in works flawlessly and as i have seen in many other discussions, I believe it is still not possible to assign privileges based on the group/roles info coming from keycloak.
The only way is to use HTTP Header Authentication (RemoteUserBackend) approach. Am i right?

Below is the netbox config for the benefit of others

> REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'social_core.backends.keycloak.KeycloakOAuth2'
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_DEFAULT_GROUPS = []
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
SOCIAL_AUTH_KEYCLOAK_KEY = 'netbox'
SOCIAL_AUTH_KEYCLOAK_SECRET = 'KuoDKWI___________BJJII'
SOCIAL_AUTH_KEYCLOAK_PUBLIC_KEY = 'MIIBIjANB_________0QAB'
SOCIAL_AUTH_KEYCLOAK_AUTHORIZATION_URL = 'https://accounts-test.mydomain.com/auth/realms/mydomain/protocol/openid-connect/auth'
SOCIAL_AUTH_KEYCLOAK_ACCESS_TOKEN_URL = 'https://accounts-test.mydomain.com/auth/realms/mydomain/protocol/openid-connect/token'

**Reply by candlerb** [reply]
I've not seen anyone successfully integrate Netbox with PSA in such a way that a group membership claim is respected.

**Reply by jeremystretch** [reply]
AFAICT this functionality is not supported by `python-social-auth`. Related, #11508 has been opened specifically for Azure AD.

**Reply by rakesh-p** [reply]
thanks for the replies. i am reverting back to HTTP header auth using apache/httpd as group sync is important for my workflow. Only downside of using HTTP header auth is i cannot get option to login as local accounts without disabling the Remote auth first. But this is still better (for me) than not able to assign privileges based on the group the user belongs to. thanks again

**Reply by jschewebbn** [reply]
@rakesh-p  note issue #7671, this bit me earlier this week with our keycloak integration.

**Reply by benley** [reply]
I've found a way to get groups data from keycloak.  In the keycloak client you have to create a Mapper for groups, like this:
![Screenshot Capture - 2023-04-28 - 16-48-13](https://user-images.githubusercontent.com/1432730/235251530-32641c9e-dcf1-4fda-9589-2b2ec37b5758.png)

I don't think it's necessary to enable all 3 of the ID token, access token, and userinfo options but I also don't remember which one makes it work.  Probably access token.

Then you can pull groups from the JWT response in your sso_pipeline_roles.py:

```py
from django.contrib.auth.models import Group

# this must match your actual keycloak client ID string!
CLIENT_ID = "netbox"


def set_role(response, user, backend, *args, **kwargs):
    try:
        roles = response['resource_access'][CLIENT_ID]['roles']
    except KeyError:
        roles = []
    user.is_staff = 'admin' in roles
    user.is_superuser = 'superuser' in roles
    user.save()


def set_groups(response, user, backend, *args, **kwargs):
    django_groups = Group.objects.all()
    oidc_groups = response.get('groups', [])
    for group in django_groups:
        try:
            if group.name in oidc_groups:
                group.user_set.add(user)
            else:
                group.user_set.remove(user)
        except Group.DoesNotExist:
            continue
```

...and add an entry to SOCIAL_AUTH_PIPELINE for `netbox.sso_pipeline_roles.set_groups` to invoke that new function.

I've chosen to use keycloak roles only for the django admin and netbox superuser roles, and group mappings come only from actual groups data.  It only syncs groups with matching names that already exist in netbox, just like the set_role function from the docs was doing with role names.

There's a good chance I've forgotten about some detail of what I did in keycloak settings, so let me know if you run into trouble with it.

**Reply by benley** [reply]
Oh also my method only syncs memberships for individual users at the time they login.  It cannot get the full membership list of a group, nor can it get the list of all groups that exist.

I think there is actually a way to do all that but it will require some extra trickery outside of the social auth pipeline.

**Comment by jcollie** [comment]
I'm wondering if switching to [django-allauth](https://github.com/pennersr/django-allauth) would work better for authenticating with external providers. Not sure how much work it would be to switch, or if it's even possible without changes to the core.

**Comment by jschewebbn** [comment]
Having gone through the work of setting up OIDC with apache and using the remote auth code, I'm wondering what would be the advantage to me switching to the python-social implementation?

**Comment by mtinberg** [comment]
I don't know all what benefits integrating authentication directly to the netbox app server brings, it works better for some people in some deployment scenerios, but for me the on-ramp to configuring the front-end reverse-proxy webserver with SSO was way easier (we already do this for other apps) and I prefer the peace of mind knowing that not a single line of (my own or 3rdparty) app code runs before authn/authz are complete, so the only exposure is to authorized users.  Which method is better probably depends on the auth provider and what they best support.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by candlerb** [reply]
I strongly agree with @mtinberg here. Django itself is a huge surface to expose, even if you only *intend* the login page to be accessible, and python-social-auth is a hairy pile of code. Having a robust and independent OIDC implementation in the proxy helps me sleep at night.

**Comment by rakesh-p** [comment]
@jschewebbn  I exactly had the same question after i made the HTTP header auth approach working and started to look into PSA. Without any detailed analysis, I assumed HTTP Header Auth introduces an additional hop (apache server in front of gunicorn) and an additional component/config to manage.   But reading this page https://demo.netbox.dev/static/docs/installation/5-http-server/ on how to secure netbox installation using HTTPS, i understand that a web server at the front is unavoidable. So I have decided to take the HTTP header auth approach and ignore PSA. Thanks everyone for your valuable inputs.

**Comment by jschewebbn** [comment]
@rakesh-p  based on my experience and the comments here, I'm going the same route.

**Comment by mtinberg** [comment]
I think of it less as an additional hop and more of a separation of concerns, let the app server deal with app technology (templates, business logic, db connections) and the web server deal with the network protocols (ssl/tls, high-performance socket handling, http protocol, virtualhosting and security) rather than make the app framework grow all those features, it really only needs a basic http service for testing and as a proxy target.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by rakesh-p** [reply]
@mtinberg  Thanks for the clear explanation. You are absolutely right!!

**Comment by rakesh-p** [comment]
I was wrong when i said above "_Only downside of using HTTP header auth is i cannot get option to login as local accounts without disabling the Remote auth first._" There is way to access the local accounts and also use OIDC at the same time. I have netbox running on HTTP but at port 8080 and apache httpd server with OIDC config serves netbox at port 443 (HTTPS). So when I hit https://mynetbox.com, I reach OIDC endpoint (keycloak in this case). When I hit, http://mynetbox.com:8080, I can use local account..  :)

**Comment by mtinberg** [comment]
That's a good idea, combined with another technique I use when the app server only listens on localhost or is blocked from remote access by firewall rules, ssh with port-forwarding or SOCKS proxying can fill that gap.

eg. ssh netbox.example.com -D 1080

with a proxy auto-config (PAC) file on a webserver you have access to

# https://netbox.example.com/static/netbox.pac
function FindProxyForURL(url, host) {
  if (host == "netbox.example.com") {
     return "SOCKS5 localhost:1080";
  } else {
     return "";
  }
}

then configure your browser proxy auto-config to that url and you'll tunnel all your netbox web traffic over SSH but nothing else, so the browser traffic is local to the target server.  That works if your organization isn't using web proxies for outbound access control, but you often can have more than one browser installed (Firefox/Chrome/Edge/Safari) so you can use a different browser with the proxy settings without breaking your normal web access.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Comment by aprudnev** [comment]
I just set up full integration (with PingID/PingOne), using this approach
- apache2 proxy runs openid connect module and makes OIDC authentication
      apt install libapache2-mod-auth-openidc

- it sends claims by HTTP headers

- netbox is configured to sync users and groups

- groups are created and configured in advance (in netbox)

Notice that python module change all header names to CAPITAL and adds HTTP_ in front.

Having this, netbox config looks this way

```
REMOTE_AUTH_ENABLED = True
#REMOTE_AUTH_ENABLED = False
REMOTE_AUTH_BACKEND = 'netbox.authentication.RemoteUserBackend'
REMOTE_AUTH_HEADER = 'HTTP_REMOTE_USER'
REMOTE_AUTH_AUTO_CREATE_USER = True
REMOTE_AUTH_DEFAULT_GROUPS = ['netbox_cvops_r']
REMOTE_AUTH_DEFAULT_PERMISSIONS = {}
REMOTE_AUTH_GROUP_HEADER = 'HTTP_OIDC_CLAIM_GROUPS'
REMOTE_AUTH_GROUP_SYNC_ENABLED = True
REMOTE_AUTH_GROUP_SEPARATOR = ','
REMOTE_AUTH_STAFF_GROUPS = ['netbox_cvops_rw']
REMOTE_AUTH_SUPERUSER_GROUPS = ['netbox_cvops_admin']
REMOTE_AUTH_USER_FIRST_NAME = 'HTTP_OIDC_CLAIM_GIVEN_NAME'
REMOTE_AUTH_USER_LAST_NAME = 'HTTP_OIDC_CLAIM_FAMILY_NAME'
REMOTE_AUTH_USER_EMAIL = 'HTTP_OIDC_CLAIM_EMAIL'

```
In addition, I enable logging and turn on debug

```
import logging.handlers
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'normal': {
            'format': '%(asctime)s %(name)s %(levelname)s: %(message)s'
        },
    },
    'handlers': {
        'file': {
            'level': 'DEBUG',
            'class': 'logging.handlers.WatchedFileHandler',
            'filename': '/var/log/netbox/netbox.log',
            'formatter': 'normal',
        },
    },
    'loggers': {
        'django': {
            'handlers': ['file'],
            'level': 'INFO',
        },
        'netbox': {
            'handlers': ['file'],
            'level': 'INFO',
        },
    },
}
```

(I _CAN_ turn on debug, it is already disabled)

And then, patch one more file to see metadata (so if I enable debug I see all headers, how they come to netbox)

```
*** middleware.py       Tue Apr 30 15:14:36 2024
--- middleware.py.ORIG  Wed May  1 17:24:22 2024
***************
*** 108,115 ****
          if not settings.REMOTE_AUTH_ENABLED:
              return
          # AuthenticationMiddleware is required so that request.user exists.
-         from pprint import pformat
-         logger.debug(pformat(request.META))
          if not hasattr(request, 'user'):
              raise ImproperlyConfigured(
                  "The Django remote user auth middleware requires the"
--- 108,113 ----

```
Then, I need to reconfigure LOGOUT so it really LOG you OUT. I use /redirect_uri OAUTH2 redirect url so I configure it here:

LOGOUT_REDIRECT_URL = '/redirect_uri?logout=%2Flogout.html'

(If you do not do it, user can not logout to change user-s id, for example - he can use incognito window etc, through).

Works well. OIDC, OAUTH2 modules (ones documented in netbox docs) looks as a joke (OIDC for example REQUIRE ISSUER or METADATA URL, CLIENT ID and SHARED KEY, and nothing is supported, and so on - so I decided that it do not make any sense to try - if someone, instead of full scale OIDC module created 'okta oidc' | 'MS oidc' then likely no one follow standards). SAML module looks reasonable (at least it contains mapping, metafile and so on) but SAML is terrible, over-blown protocol, and I skip it as much as I can. 

Apache front end is standard for OIDC with reverse proxy and contains this (be careful, if user has too many groups, it may create a very big header which may violate http standard)

```
<VirtualHost *:443>
    ProxyPreserveHost On

    # CHANGE THIS TO YOUR SERVER'S NAME
    ServerName {server name}

    SSLEngine on

   SSLCertificateFile /etc/apache2/CERT/server.crt
   SSLCertificateKeyFile /etc/apache2/CERT/server.key
   SSLCertificateChainFile /etc/apache2/CERT/chain.crt

    Alias /static /opt/netbox/netbox/static

    <Directory /opt/netbox/netbox/static>
        Options Indexes FollowSymLinks MultiViews
        AllowOverride None
        Require all granted
    </Directory>

    <Location /static>
        ProxyPass !
    </Location>

    RequestHeader set "X-Forwarded-Proto" expr=%{REQUEST_SCHEME}
    ProxyPass / http://127.0.0.1:8001/
    ProxyPassReverse / http://127.0.0.1:8001/

  # Note that OIDC Providers (OP) don't guarantee the uniqueness of these claims
    Include "/etc/httpd/conf.d/MFA_netbox.inc"
    #OIDCRemoteUserClaim email
    OIDCRemoteUserClaim sub
    OIDCPassUserInfoAs claims
    #OIDCPassClaimsAs headers
    OIDCPassClaimsAs both
    RequestHeader unset REMOTE_USER
    RequestHeader set REMOTE_USER expr=%{REMOTE_USER}

    <Location />
        AuthType openid-connect
        # We manage access on IdP side.
        Require valid-user
        #Require claim groups:Employees     # or: Require valid-user
        AuthzSendForbiddenOnFailure on
    </Location>

   Alias /logout.html /var/www/logout.html
   <Location /logout.html>
        ProxyPass !
        Require all granted
    </Location>
    <Location /api/>
        <RequireAny>
            AuthType openid-connect
            Require valid-user
            # Allow access to /api/ with token instead of OIDC:
            # https://httpd.apache.org/docs/2.4/howto/access.html#env
            Require expr "%{HTTP:Authorization} =~ /^API-access-token /"
        </RequireAny>
        AuthzSendForbiddenOnFailure on
    </Location>
</VirtualHost>
```

and OIDC include



 ```
# cat MFA_netbox.inc
OIDCProviderMetadataURL https://{meta issuer URL}/.well-known/openid-configuration
 OIDCClientID {client ID}
 OIDCClientSecret {client secret}
 OIDCHTMLErrorTemplate /var/www/error_prodping.html
 OIDCStateMaxNumberOfCookies 10 true
 OIDCCryptoPassphrase {temporary-key}
 OIDCRedirectURI /redirect_uri
 OIDCScope "openid email profile groups"
<Location /redirect_uri>
  AuthType openid-connect
  Require valid-user
</Location>
ProxyPass /redirect_uri !


```

**Reply by vishnubraj** [reply]
@aprudnev @candlerb  What should be the `redirect url` and `login url` configured on the OIDC provider? 
I am using the below values, whenever I click on the app created in the OIDC provider console, it again goes back to the console page, not redirecting to netbox link. 
```
redirect url: https://netbox.ABC.com/oauth2callback
login url: https://netbox.ABC.com/
```

**Reply by candlerb** [reply]
If you mean with the django-social-auth integration, I believe the callback URL is something like

```
https://netbox.ABC.com/oauth/complete/foo/
```

where "foo" is the social-auth backend module you're using. But you're supposed to figure this out for yourself.

**Reply by vishnubraj** [reply]
Thanks @candlerb  I am using the apache oidc module so the below one worked. 

```
redirect url: https://netbox.ABC.com/oauth2callback
```

**Comment by candlerb** [comment]
> Notice that python module change all header names to CAPITAL and adds HTTP_ in front.

And it converts dash to underscore.

Note that Gunicorn 22.0.0 (which Netbox 3.7.6 picks up) has stopped accepting header names containing underscores by default, so the best thing to do now is to use a dash in the header:

```
    RequestHeader unset Remote-User
    RequestHeader set Remote-User expr=%{REMOTE_USER}
```

After mapping, this will still give you `HTTP_REMOTE_USER` in Netbox. See #15811

> be careful, if user has too many groups, it may create a very big header which may violate http standard

I don't believe the HTTP standard specifies any limit. However, various implementations have limits for the length of a single header and/or the total size of all headers, which may or may not be configurable. For example, Apache has [LimitRequestFieldSize](https://httpd.apache.org/docs/2.4/mod/core.html#limitrequestfieldsize).

**Reply by vishnubraj** [reply]
in addition to the above,  for anyone trying Apache OIDC,  use the below config in the httpd config file to change the OIDC claim prefix, without that Gunicorn will not accept the OIDC claims sent by the web server. by default apache sends the OIDC claims with underscore which is not accepted in Gunicorn. 


`OIDCClaimPrefix OIDC-Claim-`

**Reply by jschewebbn** [reply]
Make sure you update your netbox configuration for the new header variables as well. An alternate solution is to add "header_map = 'dangerous'" to your gunicorn.py config file.

**Comment by aprudnev** [comment]
I checked OIDC provider, and I did not like what I see. SO I instead set up
OIDC on reverse proxy and passed user etc to the application thru it; it
works great (and apache OIDC module was designed by PingID engineers, so it
is fully up to the standards).
